### PR TITLE
planner: fix the issue that the optimizer cannot convert OUTER JOIN to INNER JOIN with nested AND/OR in some cases

### LINKED ISSUE #49616: planner: the optimizer cannot convert OUTER JOIN to INNER JOIN with nested AND/OR predicates in some cases
## Enhancement
```
create table t1 (k int, a int);
create table t2 (k int, b int, key(k));

explain select /*+ tidb_inlj(t2, t1) */ *
  from t2 left join t1 
    on t1.k=t2.k
  where 
    a>0 or (a=0 and b>0);
+--------------------------------+----------+-----------+---------------+---------------------------------------------------------------+
| id                             | estRows  | task      | access object | operator info                                                 |
+--------------------------------+----------+-----------+---------------+---------------------------------------------------------------+
| Selection_7                    | 9990.00  | root      |               | or(gt(test.t1.a, 0), and(eq(test.t1.a, 0), gt(test.t2.b, 0))) |
| └─HashJoin_8                   | 12487.50 | root      |               | left outer join, equal:[eq(test.t2.k, test.t1.k)]             |
|   ├─TableReader_14(Build)      | 9990.00  | root      |               | data:Selection_13                                             |
|   │ └─Selection_13             | 9990.00  | cop[tikv] |               | not(isnull(test.t1.k))                                        |
|   │   └─TableFullScan_12       | 10000.00 | cop[tikv] | table:t1      | keep order:false, stats:pseudo                                |
|   └─TableReader_11(Probe)      | 10000.00 | root      |               | data:TableFullScan_10                                         |
|     └─TableFullScan_10         | 10000.00 | cop[tikv] | table:t2      | keep order:false, stats:pseudo                                |
+--------------------------------+----------+-----------+---------------+---------------------------------------------------------------+
```

In the case above, the `tidb_inlj` hint doesn't take effect, because the optimizer cannot convert this OUTER JOIN into an INNER JOIN, since the predicate `a>0 or (a=0 and b>0)` doesn't pass the null-rejective check (see `isNullRejected`, `evaluateExprWithNullInNullRejectCheck`). 
But actually, the predicate is null-rejective since no matter the value of `b`, if `a` is `null`, then the eventual value of this predicate is `null`, and so all unmatched OUTER JOIN rows (whose `a` is `null`) will be eliminated by this predicate. So it's safe to convert this OUTER JOIN to an INNER JOIN, but the optimizer didn't recognize this.

This issue is caused by https://github.com/pingcap/tidb/pull/38430, our current null-rejective check is too strict.

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve?

Issue Number: close #49616

Problem Summary: planner: fix the issue that the optimizer cannot convert OUTER JOIN to INNER JOIN with nested AND/OR in some cases

### What changed and how does it work?

planner: fix the issue that the optimizer cannot convert OUTER JOIN to INNER JOIN with nested AND/OR in some cases

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/planner/core/integration_test.go [hunk 1]
```diff
@@ -1626,6 +1626,17 @@ func TestLeastGretestStringPushDownToTiFlash(t *testing.T) {
...
 
+func TestIssue49616(t *testing.T) {
+	store := testkit.CreateMockStore(t)
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec(`use test`)
+	tk.MustExec(`create table t1 (k int, a int)`)
+	tk.MustExec(`create table t2 (k int, b int, key(k))`)
+	tk.MustHavePlan(`select /*+ tidb_inlj(t2, t1) */ *
+  from t2 left join t1 on t1.k=t2.k
+  where a>0 or (a=0 and b>0)`, `IndexJoin`)
```
Comment (hawkingrei) [inline pkg/planner/core/integration_test.go:1637]: I suggest to move this test into ```pkg/planner/core/issuetest```.

File: pkg/planner/core/rule_predicate_push_down.go [hunk 2]
```diff
@@ -454,6 +458,47 @@ func isNullRejected(ctx sessionctx.Context, schema *expression.Schema, expr expr
...
+	}
+	for i := 0; i < 2; i++ {
+		andFunc := isFunc(orFunc.GetArgs()[i], ast.LogicAnd)
+		if andFunc == nil {
+			continue
+		}
+		if !isNullRejected(ctx, schema, orFunc.GetArgs()[1-i]) {
+			continue // the other side should be null-rejected: null-rejected OR (... AND ...)
+		}
+		for _, andItem := range expression.FlattenCNFConditions(andFunc) {
```
Comment (time-and-fate) [inline pkg/planner/core/rule_predicate_push_down.go:493]: I think it should be `SplitCNFItems` here.
```suggestion
		for _, andItem := range expression.SplitCNFItems(andFunc) {
```

File: pkg/planner/core/integration_test.go (hunk 1 above)
Comment (qw4990) [reply pkg/planner/core/integration_test.go]: fixed

**Review by time-and-fate — APPROVED** [review]

**Review by winoros — APPROVED** [review]

**Review by AilinKid — APPROVED** [review]

**Review by ghazalfamilyusa — COMMENTED** [review]
The general solution to null filtering condition NFC is simple and wondering why we need to limit to this special case. 
NFC (A and B) = NFC(A) OR NFC(B)
NFC (A OR B) = NFC(A) AND NFC(B) 
We do have NFC logic for primitive conditions and should be easy to do this general fix.

**Pull request merged**